### PR TITLE
Calico-4373: default missing protocol to TCP; don't let single port overwrite all ports

### PR DESCRIPTION
**Pull request by mattfenwick**
## Description

Bug fix: fixes https://github.com/projectcalico/calico/issues/4373.  There were a couple problems with port parsing that this addresses:

 - missing protocols were not treated as TCP, contradicting the network policy spec
 - allowing a specific port for a protocol could cause an 'allow all ports' for a protocol to be lost

Test plan:

```
libcalico-go/lib/backend/k8s/conversion$ go test .

libcalico-go$ sudo make test
```

## Todos
- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
When interpreting Kubernetes NetworkPolicy ports, Calico now interprets an empty port struct as "all TCP" as per the NetworkPolicy spec.  Previously, empty structs were ignored. 
```

### CODE REVIEW

File: lib/backend/k8s/conversion/conversion.go [hunk 1]
```diff
@@ -400,6 +400,9 @@ func (c converter) k8sRuleToCalico(rPeers []networkingv1.NetworkPolicyPeer, rPor
 		if p.Protocol != nil {
 			protval := kapiv1.Protocol(fmt.Sprintf("%s", *p.Protocol))
 			port.Protocol = &protval
+		} else {
```
Comment (fasaxc) [inline lib/backend/k8s/conversion/conversion.go:403]: I think this makes line 398 redundant so you can remove that line and its comment

File: lib/backend/k8s/conversion/conversion.go [hunk 2]
```diff
@@ -430,7 +433,14 @@ func (c converter) k8sRuleToCalico(rPeers []networkingv1.NetworkPolicyPeer, rPor
 		}
 
 		pStr := protocol.String()
-		protocolPorts[pStr] = append(protocolPorts[pStr], calicoPorts...)
+		// treat nil as 'all ports'
+		if calicoPorts == nil {
```
Comment (fasaxc) [inline lib/backend/k8s/conversion/conversion.go:437]: Comment above on line 426 is now out of date; `calicoPorts` can be nil, now.

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looking good, just a couple of comment-level nits.

File: lib/backend/k8s/conversion/conversion.go (hunk 1 above)
Comment (mattfenwick) [reply lib/backend/k8s/conversion/conversion.go:398]: 👍

File: lib/backend/k8s/conversion/conversion.go (hunk 2 above)
Comment (mattfenwick) [reply lib/backend/k8s/conversion/conversion.go:433]: 👍

**Review by fasaxc — APPROVED** [review]

**Pull request merged**